Refactor argument parsing and passing to a json-based approach instead of code injection

rcannood opened the pull request:
## Describe your changes

This PR refactors Viash to use JSON-based parameter passing instead of direct code injection. This makes argument value handling more robust, easier to debug, and correctly handles special characters like backticks, quotes, dollar signs, and newlines.

## Related issue(s)

Fixes #375, #619, #705, #763, #821, #840

## Changes

- **New `io.viash.languages` package**: Language trait with implementations for Bash, Python, R, JavaScript, Scala, C#, and Nextflow
- **JSON-based parameter passing**: Arguments are stored in `params.json` and parsed at runtime using language-specific JSON parsers (no external dependencies required)
- **Build-time code injection**: JSON parsing code is injected at build time, not runtime


## Breaking changes

### Multiple-value input arguments are now arrays

**Before:**
```bash
# Arguments with multiple: true were semicolon-separated strings
IFS=';' read -ra items <<< "$par_inputs"
for item in "${items[@]}"; do
  echo "$item"
done
```

**After:**
```bash
# Arguments with multiple: true are now bash arrays
for item in "${par_inputs[@]}"; do
  echo "$item"
done
```

## JSON Structure

```json
{
  "par": {
    "input": "/path/to/input.txt",
    "whole_number": 42,
    "real_number": 3.14,
    "multiple": ["a", "b", "c"]
  },
  "meta": {
    "name": "component_name",
    "resources_dir": "/path/to/resources",
    "executable": "/path/to/executable",
    "config": "/path/to/.config.vsh.yaml",
    "temp_dir": "/tmp",
    "cpus": 2,
    "memory_b": 2000000000
  },
  "dep": {
    "dependency_name": "/path/to/dependency"
  }
}
```

## Reviewing the viash work directory

To inspect the work directory and `params.json` file during execution:

```bash
# Build a test component
bin/viash build src/test/resources/test_languages/bash/config.vsh.yaml -o /tmp/test_component

# Run with VIASH_KEEP_WORK_DIR to preserve the work directory
VIASH_KEEP_WORK_DIR=true /tmp/test_component/test_languages_bash \
  /tmp/test_component/.config.vsh.yaml \
  --real_number 3.14 \
  --whole_number 42 \
  -s "hello world"

# The work directory location is printed; inspect params.json:
cat /tmp/viash_work_test_languages_bash_*/params.json | jq .
```

## New feature: UNDEFINED and UNDEFINED_ITEM

This PR adds proper support for explicitly unsetting argument values:

### Unsetting Single-Value Arguments

Pass the literal `UNDEFINED` (unquoted) to set an argument to null:

```bash
./my_component --optional_arg UNDEFINED
```

Result: `par["optional_arg"]` is `None` (Python), `NULL` (R), `null` (JS), or unset (Bash).

### Missing items in multi-value arguments

Use `UNDEFINED_ITEM` in semicolon-separated values to represent null elements:

```bash
./my_component --values "item1;UNDEFINED_ITEM;item3"
```

Result: `par["values"]` is `["item1", None, "item3"]` (Python) or equivalent.

### Passing literal "UNDEFINED" string

Use nested quotes to pass "UNDEFINED" as a literal string:

```bash
./my_component --arg '"UNDEFINED"'    # par["arg"] = "UNDEFINED"
./my_component --arg "'UNDEFINED'"    # par["arg"] = "UNDEFINED"
```

### Unsetting computational requirements

You can also pass UNDEFINED to unset `---cpus` or `---memory` defaults:

```bash
./my_component ---cpus "UNDEFINED" ---memory "UNDEFINED"
```

## Checklist

- [x] All tests passing (55 component + 16 Nextflow + 190 E2E)
- [x] Bash 3.2 compatibility verified
- [x] BSD/macOS compatibility verified
- [x] Breaking changes documented
- [x] Changelog updated

rcannood committed:
first commit

rcannood committed:
fix usage of wrong varible

rcannood committed:
fix missing $

rcannood committed:
fix variables again

rcannood committed:
unset first, just in case

rcannood committed:
replace `IFS=... for var in $myarg` with `for i in ${!myarg[@]}`

rcannood committed:
fix type mismatch error

rcannood committed:
move dependencies to mods

rcannood committed:
wip helper functions

rcannood committed:
rename helper functions

rcannood committed:
fixes to generated code

rcannood committed:
add more comments

rcannood committed:
remove unused code

rcannood committed:
make certain BashWrapper functions private

rcannood committed:
Merge remote-tracking branch 'origin/develop' into switch_to_arrays

rcannood committed:
improve yaml rendering

rcannood committed:
fixes

rcannood committed:
more fixes

rcannood committed:
only input multiples are arrays

rcannood commented:
Status update:

What needs to happen in this PR:

* [ ] BashWrapper
  - [x] Passed argument values get stored in a bash array instead of a string-delimited variable
  - [ ] While doing the argument parsing, we need to process quoted strings and any escaped values. This also means that an unquoted `UNDEFINED` needs to be stored in a way that allows us to know later on that the argument was unset and not that the user actually provided a quoted `"UNDEFINED"` value instead. Same for `UNDEFINED_ITEM`. I would be ok with just storing something like `@@VIASH_UNDEFINED@@` and `@@VIASH_UNDEFINED_ITEM@@`.
  - [ ] Argument values are translated into a yaml to pass to the script (partially done)
  - [ ] The script can now also be stored in the same temp dir as where the yaml resides, since we don't need to do any code injection
  - [ ] See `Proposed flow of parsing in BashWrapper`
* [ ] ExecutableRunner:
  - [x] Ensure that any time we generate code that looks at the VIASH_PAR, that this is able to handle arrays
  - [ ] Also handle values set to `@@VIASH_UNDEFINED@@` or values set to `@@VIASH_UNDEFINED_ITEM@@`
* [ ] NextflowRunner: Needs to be updated to also create a yaml from the inputs it receives, and store the script in the same tempdir since we don't need to do any script injection
* [ ] Script: All scripting languages need to be updated such that they don't generate dictionaries from the environment variables, but instead are able to parse YAML using the base functions in that scripting language. While we're at it, I would create an `io.viash.language` or something like that and add all of the supported scripting languages there.

Proposed flow of parsing in BashWrapper:
  - [x] Preparse: do nothing. Make sure the arrays don't already exist.
  - [ ] During parse:
    - [ ] If the argument is multiple false and the environment variable already exists, throw an error
    - [ ] If the argument is multiple true and the environment variable is not an array and equals `@@VIASH_UNDEFINED@@`, throw an error
    - [ ] If the argument is multiple true and the environment variable doesn't exist yet, create an array
    - [ ] If the passed value is equal to `UNDEFINED`, set the environment variable to `@@VIASH_UNDEFINED@@`. (Can we later still check if the variable is an array or not?)
    - [ ] If the passed value is equal to `UNDEFINED_ITEM`, change it into `@@VIASH_UNDEFINED_ITEM@@`
    - [ ] Resolve quotes and escaped characters
    - [ ] If the argument is multiple true, add the value to the array, else set the environment variable to the value
  - [ ] Postparse:
    - [ ] If the variable equals `@@VIASH_UNDEFINED@@`, unset the variable.
    - [ ] Else, if the variable is unset and the argument has a default, set the variable to the default values. Ensure that defaults are stored as an array.
  - [ ] Prerun:
    - [x] Convert argument values into a yaml, which contains a `{par: {...}, meta: {...}, dependencies: {...}}`.
    - [ ] Ensure that the format of the yaml is as follows:
      ```yaml
      par:
        arg1: "value1"
        arg2: "value2"
        arg_multiple:
          - "foo"
          - "bar"
          - null # an UNDEFINED_ITEM was passed
        arg_with_default: null # an UNDEFINED was passed
      meta:
        # etc...
      ```
    - [ ] ↑ This should allow for easy processing with awk. Notes:
      - [x] That is, each argument gets its own row
      - [ ] If the argument is multiple true, each item also gets its own row
      - [ ] Strings are always escaped
      - [ ] `@@VIASH_UNDEFINED@@` is translated into `null`
      - [ ] `@@VIASH_UNDEFINED_ITEM@@` in `[..., null, ...]`.

rcannood committed:
Merge remote-tracking branch 'origin/develop' into switch_to_arrays

rcannood committed:
wip add tests

rcannood committed:
implement escaping en quoting during bashwrapper parsing

rcannood committed:
Merge remote-tracking branch 'origin/develop_0_9' into switch_to_arrays

rcannood committed:
fix renderyaml function

rcannood committed:
add docs

rcannood committed:
unset variables if they contain @@VIASH_UNDEFINED@@

rcannood committed:
Merge branch 'develop' into switch_to_arrays

rcannood committed:
wip python render attempt

rcannood committed:
create languages with viash yaml parsers

rcannood committed:
add a unit test

rcannood committed:
restructure code

rcannood committed:
fix parsers, add tests

rcannood committed:
fix tests and parsers

rcannood committed:
try to create script in local tempdir

rcannood committed:
switch to json parsers

rcannood committed:
add tests

rcannood committed:
refactor languages <> scripts

rcannood committed:
fix run cmd in bashwrapper

rcannood committed:
fix viash json rendering indentation

rcannood committed:
fix scala parse json test

rcannood committed:
add json parser test suite

rcannood committed:
Merge remote-tracking branch 'origin/main' into switch_to_arrays

rcannood committed:
fix nextflow helper

rcannood committed:
wip extra test

rcannood committed:
refactor script generation

rcannood committed:
fix set wd to resources

rcannood committed:
dot is now included in the extension

rcannood committed:
add progress file

rcannood committed:
move injectionmods code

rcannood committed:
fix r component

rcannood committed:
bump minimum docker version to 4.4 (due to declare -g)

rcannood committed:
fix bash json parser

rcannood committed:
fix bash json parser test

rcannood committed:
improve unit test to get more info on what failed

rcannood committed:
fix bash component

rcannood committed:
add cleanup registry to make sure multiple trap calls don't override one another

rcannood committed:
remove outdated code

rcannood committed:
add tests for the viashcleanupregistry

rcannood committed:
automate running tests

rcannood committed:
fix scala parsers

rcannood committed:
fix component

rcannood committed:
use anonymous class

rcannood committed:
fix javascript component and code

rcannood committed:
add override

rcannood committed:
update progress

rcannood committed:
fix executable

rcannood committed:
Wip refactor nextflow runner

rcannood committed:
fix tests

rcannood committed:
fix quoting when checking choices

rcannood committed:
add touch command

rcannood committed:
fix tests

rcannood committed:
refactor viash config inject

rcannood committed:
fix testbash component

rcannood committed:
fix test

rcannood committed:
reduce verbosity

rcannood committed:
fix tests

rcannood committed:
fix test

rcannood committed:
fix mount detection

rcannood committed:
fix warning message

rcannood committed:
fix test

rcannood committed:
fix description

rcannood committed:
update viash par

rcannood committed:
fix tests

rcannood committed:
fix formatting

rcannood committed:
fix trap

rcannood committed:
try to fix missing scala and dotnet dependencies

rcannood committed:
try to make bash 3.2 compatible

rcannood committed:
switch back to bash 3.2

rcannood committed:
fix error message

rcannood committed:
fix test

rcannood committed:
fix escaping

rcannood committed:
extgend test

rcannood committed:
add tests for renderjson function

rcannood committed:
adjust helper functions, expand tests

rcannood committed:
expand tests; fix summary

rcannood committed:
fixes for viashrenderjson

rcannood committed:
fix test

rcannood committed:
add more tests

rcannood committed:
add additional tests

rcannood committed:
add test

rcannood committed:
add bashutil test

rcannood committed:
refactor bash util test suite

rcannood committed:
fix viashparsejson

rcannood committed:
remove debug notes

rcannood committed:
update test

rcannood committed:
store work dir inside current temp dir for ease of debugging

rcannood committed:
resolve issue with temp dir on mac os x

rcannood committed:
try to fix viash parse argument with native bash instead of awk

rcannood committed:
refactor example code

rcannood committed:
add entry to changelog

rcannood committed:
update intermediate progress file

rcannood committed:
remove intermediate progress file

rcannood committed:
Merge remote-tracking branch 'origin/main' into switch_to_arrays

rcannood committed:
fix edge cases

rcannood committed:
fix test

rcannood committed:
update docs

rcannood committed:
update changelog

rcannood committed:
unset memory and cpus with UNDEFINED

rcannood committed:
satisfy linter

tverbeiren committed:
Add nasty_val to break test

rcannood commented on src/main/resources/io/viash/languages/bash/ViashParseJson.sh:
```diff
@@ -0,0 +1,295 @@
+      nested_json+="$line"
+      
+      # Count braces to track when nested object ends
+      if [[ "$line" =~ \{ ]]; then
```
Oh, thanks for spotting this!

rcannood committed:
con't wrap testexecutable as nextflow

rcannood committed:
fix test

rcannood committed:
fix renderjson function

rcannood committed:
add test to verify fix

rcannood committed:
validate unquoted values

rcannood committed:
Merge branch 'switch_to_arrays' of github.com:viash-io/viash into switch_to_arrays

hcannoodt committed:
fix curly brackets in strings for bash json parser (#868)

rcannood committed:
add improved json parser

rcannood committed:
improve speed

Commit:
Improved bash json parser (#869)

* add improved json parser

* improve speed

rcannood committed:
improve parser

rcannood committed:
Implement hybrid approach

rcannood committed:
add argument

rcannood committed:
Update tests

rcannood committed:
Merge branch 'switch_to_arrays' of github.com:viash-io/viash into switch_to_arrays

rcannood committed:
move regular json parser to compatibility; add new jq parser

rcannood committed:
add argument to bash script

rcannood committed:
simplify code

rcannood committed:
update tests

tverbeiren commented:
> Updates to mitigate breaking changes:
> 
>     * `jsonlite` is a requirement for R components
>       
>       * [ ]  Write custom parser
>       * [ ]  If installed, use `jsonlite` else use custom parser (with deprecation warning)
>       * [ ]  Add unit tests for both situation

This is nicely covered, including tests!

>     * The same for `jq` in `bash`

I don't see how this is covered?

>     * `bash` uses arrays: All components that use bash will break !
>       * Add flag for backward compatibility → with deprecation warning

I don't see where this is covered either?

rcannood committed:
fix imports

rcannood committed:
extend unit tests

rcannood committed:
update components

rcannood committed:
add tests

rcannood committed:
unset arguments before assignment

rcannood committed:
add jq as dependency

rcannood committed:
fix tests

rcannood committed:
fix tests

hcannoodt committed:
update old image definitions for js and r (#872)

the old images are on old debian versions that aren't maintained any longer, resulting in unmaintained repositories where we can't update or install jq any longer

hcannoodt committed:
Merge branch 'main' into switch_to_arrays

hcannoodt committed:
fix scala json test

Use Scala 3.6 as Scala 3.8 no longer supports Java 11
Start Scala/Bloop engine as a bunch of downloads need to happen and needs some time to wake up preventing possible stuck servers down the road

hcannoodt committed:
Merge branch 'main' into switch_to_arrays
